src/python/WMCore/Services/Rucio/Rucio.py

```diff
@@ -229,3 +239,246 @@ def getPFN(self, nodes=None, lfns=None, destination=None, protocol='srmv2', cust
+            response = self.attachDIDs(kwargs.get('rse'), container, name, scope)
+        return response
+
+    def attachDIDs(self, rse, superDID, dids, scope='cms'):
+        """
+        _attachDIDs_
+
+        Create a list of files - in bulk - to a RSE. Then attach them to the block name.
+        :param rse: string with the RSE name
+        :param superDID: upper structure level (can be a container or block. If attaching blocks,
+             then it's a container name; if attaching files, then it's a block name)
+        :param dids: either a string or a list of data identifiers (can be block or files)
+        :param scope: string with the scope name
+        :return: a boolean to represent whether it succeeded or not
+        """
+        if isinstance(dids, basestring):
+            dids = [dids]
+        dids = [{'scope': scope, 'name': did} for did in dids]
+
+        response = False
+        try:
+            response = self.cli.attach_dids(scope, superDID, dids=dids, rse=rse)
+        except DuplicateContent:
+            self.logger.warning("Dids: %s already attached to: %s", dids, superDID)
+            response = True
+        except FileAlreadyExists:
+            self.logger.warning("Failed to attach files already existent on block: %s", superDID)
+            response = True
+        except DataIdentifierNotFound:
+            self.logger.error("Failed to attach dids: %s. Parent DID %s does not exist.", dids, superDID)
+        except Exception as ex:
+            self.logger.error("Exception attaching dids: %s to: %s. Error: %s",
+                              dids, superDID, str(ex))
+        return response
+
+    def createReplicas(self, rse, files, block, scope='cms', ignoreAvailability=True):
+        """
+        _createReplicas_
+
+        Create a list of files - in bulk - to a RSE. Then attach them to the block name.
+        :param rse: string with the RSE name
+        :param files: list of dictionaries with the file names and some of its meta data.
+            E.g.: {'name': lfn, 'bytes': size, 'scope': scope, 'adler32': checksum, 'state': 'A'}
+            State 'A' means that the replica is available
+        :param block: string with the block name
+        :param scope: string with the scope name
+        :param ignore_availability: boolean to ignore the RSE blacklisting
+        :return: a boolean to represent whether it succeeded or not
+        """
+        if isinstance(files, dict):
+            files = [files]
+        for item in files:
+            item['scope'] = scope
+
+        # TODO: test to make sure 'state' is a valid argument
+        response = False
+        try:
+            # add_replicas(rse, files, ignore_availability=True)
+            response = self.cli.add_replicas(rse, files, ignoreAvailability)
+        except Exception as ex:
+            self.logger.error("Failed to add replicas for: %s and block: %s. Error: %s", files, block, str(ex))
+
+        if response:
```

Comment:
I'm not sure the logic here makes sense to try to attach DIDs every time a replica is made. 

In general, the way I'd think about it is 
- tell Rucio which DIDs exist and how they are related
- tell Rucio where the replicas of files are

So attaching files to block is part of the first part.

Reply:
Ok, this might be where I get confused.
When we add a new replica (which is a file at a given location), doesn't it meet your second bullet comment?

In short, we add a replica also providing their RSE availability, then we attach those replicas to a block.